### PR TITLE
feat: Introduce interfaces for metrics instrumentation

### PR DESCRIPTION
**Pull request by blakeli0**
This PR introduces new public interfaces and refactors existing interfaces for metrics instrumentation. All the interfaces are framework-agnostic. Subsequent PRs should implement interfaces with an observability framework. For example, introduce an `OpenTelemetryMetricsRecorder` that implements `MetricsRecorder`. All new methods are created with Javadocs, they are expected to be updated in the implementation PR if needed.

See go/java-gapic-otel-metrics-design for details.

### CODE REVIEW

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java
Comment (lqiu96) [inline gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java:82]: qq, having a `TRACER_KEY` in the CallOptions this a gRPC specific thing? Would it make sense to also have this for HttpJson and if so, would it make sense to make this as transport neutral as possible?

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java
Comment (lqiu96) [inline gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java:39]: Perhaps we can create a ticket for this so we don't forget about it. @ddixit14 

If I'm understanding this correctly, once stage 1 is complete, we can remove this class.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (lqiu96) [inline gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: @blakeli0 I am now having some second thoughts on the `serviceName()` getter's name. I think there might be some slight discrepancy between the intended result and the function name.

`getServiceName()` returns the name that was configured in the default_host. For example, asset's will return cloudasset: https://github.com/googleapis/google-cloud-java/blob/9785f8cfd43db0a8968f086a1461242da9d16cb5/java-asset/google-cloud-asset/src/main/java/com/google/cloud/asset/v1/stub/AssetServiceStubSettings.java#L866-L868

even though the intended result for this probably should be `asset`. 

Not something blocking this PR, but just wondering your thoughts on this.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java
Comment (lqiu96) [inline gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java:41]: Do we have a timeline of when to remove this Beta annotation? Same for the other annotations above.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java
Comment (lqiu96) [inline gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java:45]: Would it make sense to add some default Javadocs now (I think it would apply to all implementations)? Or is this meant for us to fill out later?

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (ddixit14) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: @lqiu96 why should the intended name be `asset` ?  
@blakeli0 Is there any way we could test this? that for a particular service, this is the intented serviceName ?

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java
Comment (ddixit14) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java:39]: sure, I'll create a ticket for this.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: > @lqiu96 why should the intended name be `asset` ? @blakeli0 Is there any way we could test this? that for a particular service, this is the intented serviceName ?

I don't want to say what the intended name *should* be. I think it might be a bit weird for customers since we wouldn't have consistent service names.

java-asset's serviceName is cloudasset (link above)
java-biglake's serviceName is biglake (https://github.com/googleapis/google-cloud-java/blob/458516b462e0a4494f32815fe9d0e6f0b30353f1/java-biglake/google-cloud-biglake/src/main/java/com/google/cloud/bigquery/biglake/v1/stub/MetastoreServiceStubSettings.java#L409-L411)

I think most services would have the service name of java-{serviceName} except for asset and possibly a few others.

It might not be a big concern after all, but something just doesn't sit right with me knowing that serviceName is pulled from a URI (default_host value) and not the package name or some config value (name_pretty or api name).

Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: FYI, your link doesn't work for me. Seems like there is no defined `serviceName` and we've somehow settled on the same definition 😆 

Not blocking this PR, but it would be great if there was an official definition for serviceName. Perhaps a question for the core team.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java:41]: I'm expecting them to be removed when we GA Otel Metrics, maybe later this year. The deliverables in Q1 is only for Spanner hence I consider them as a private preview feature.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: @lqiu96 I think the term `serviceName` you introduced still makes sense, because we call the yaml and json config for each versioned protos [service yaml](https://github.com/googleapis/googleapis/blob/master/google/cloud/asset/v1/cloudasset_v1.yaml) and [service config](https://github.com/googleapis/googleapis/blob/master/google/cloud/asset/v1/cloudasset_grpc_service_config.json) respectively. we may want to add more docs to explain that it is used for constructing service endpoints.  
@ddixit14 I would call `java-{serviceName}` a library name or repo name, as it could includes multiple versioned services or admin services.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/BaseApiTracer.java:39]: I'm planning to deprecate `BaseApiTracer` in a follow up PR, it is an `InternalApi` and only used by Bigtable outside  of our team. So we can could deprecate it anytime want, and remove it after Bigtable removes the usage of it.

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MetricsRecorder.java]: Added

File: gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java
Comment (blakeli0) [reply gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java:82]: Yes, it makes sense. Currently it is only used by [GrpcChannelUUIDInterceptor](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcChannelUUIDInterceptor.java#L47), if we have a need to use it in [HttpJsonHeaderInterceptor](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonHeaderInterceptor.java#L52), then we can add it to HttpJsonCallOptions. We can refactor it when we have the need in the future.

**Review by ddixit14 — APPROVED** [review]

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: Ok, to wrap up the conversation above: I'll add some additional docs clarifying the intended use case (constructing the endpoint) and how the this value is constructed (from the config files).

It is fine that serviceName in the library (pulled from `getServiceName()`) doesn't 100% match `java-{serviceName}`.

**Review by lqiu96 — APPROVED** [review]

File: gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/tracing/MethodName.java:46]: > It is fine that serviceName in the library (pulled from `getServiceName()`) doesn't 100% match `java-{serviceName}`.

Correct. The library/repo name may not always be in the format of `java-{serviceName}` and I would not refer `java-{serviceName}` as the format for library/repo name.

**Pull request merged**
